src/storage/compressed_tile.cpp

```diff
@@ -0,0 +1,301 @@
+
+  std::sort(column_values.begin(), column_values.end(), CompareLessThanBool);
+
+  type::Value median = column_values[num_tuples / 2];
+  std::vector<type::Value> modified_values(num_tuples);
+
+  while (true) {
+    try {
+      type::Value min_diff =
+          column_values[0].Subtract(median).CastAs(compression_type);
+      type::Value max_diff =
+          column_values[num_tuples - 1].Subtract(median).CastAs(
+              compression_type);
+      for (oid_t k = 0; k < num_tuples; k++) {
+        modified_values[k] =
+            actual_values[k].Subtract(median).CastAs(compression_type);
+      }
+
+      LOG_TRACE("Base value selected = %s", median.ToString().c_str());
+      break;
+    } catch (Exception &e) {
+      if (column_type == (compression_type + 1)) {
+        modified_values.resize(0);
+        break;
+      }
+      compression_type = static_cast<type::Type::TypeId>(
+          static_cast<int>(compression_type) + 1);
+    }
+  }
+  return modified_values;
+}
+
+void CompressedTile::InsertTuple(const oid_t tuple_offset, Tuple *tuple) {
+  if (IsCompressed()) {
+    LOG_TRACE("Peloton does not support insert into compressed tiles.");
+    PL_ASSERT(false);
+  } else {
+    Tile::InsertTuple(tuple_offset, tuple);
+  }
+}
+
+type::Value CompressedTile::GetValue(const oid_t tuple_offset,
+                                     const oid_t column_id) {
+  type::Value deserizedValue = Tile::GetValue(tuple_offset, column_id);
+
+  if (!IsCompressed()) {
+    return deserizedValue;
+  }
+
+  if (GetCompressedType(column_id) == type::Type::INVALID) {
+    return deserizedValue;
+  }
+
+  type::Type::TypeId compressed_type = GetCompressedType(column_id);
+
+  PL_ASSERT(compressed_type != type::Type::INVALID);
+
+  type::Value original = GetUncompressedValue(column_id, deserizedValue);
+
+  return original;
+}
+
+type::Value CompressedTile::GetValueFast(const oid_t tuple_offset,
+                                         const size_t column_offset,
+                                         const type::Type::TypeId column_type,
+                                         const bool is_inlined) {
+  type::Value deserizedValue =
+      Tile::GetValueFast(tuple_offset, column_offset, column_type, is_inlined);
+
+  if (!IsCompressed()) {
+    return deserizedValue;
+  }
+
+  oid_t column_id = GetColumnFromOffsest(column_offset);
+
+  PL_ASSERT(column_id < column_count);
+
+  if (GetCompressedType(column_id) != type::Type::INVALID) {
+    type::Value original = GetUncompressedValue(column_id, deserizedValue);
+    return original;
+  }
+
+  return deserizedValue;
+}
+
+void CompressedTile::SetValue(const type::Value &value,
+                              const oid_t tuple_offset, const oid_t column_id) {
+  if (IsCompressed()) {
+    if (GetCompressedType(column_id) != type::Type::INVALID) {
+      LOG_TRACE("Peloton does not support SetValue on a Compressed Class");
+      PL_ASSERT(false);
```

Comment:
Once a tile has been compressed, its values can't be modified since the new value might be out of the compression range. In this case the entire tile will have to be uncompressed at runtime (leading to poor performance). The PL_ASSERT(false) will have to stay. We will be changing the call to the compressTiles() function in the tile_group.cpp file to handle this. The current code was only used to check the correctness of the compression. (Inserts and selects)